src/audio/pipewire.c

```diff
@@ -0,0 +1,254 @@
+    buf->datas[0].chunk->offset = 0;
+    buf->datas[0].chunk->size = number_of_frames;
+    buf->datas[0].chunk->stride = stride;
+    pw_stream_queue_buffer(state->stream, b);
+}
+// pipewire internal: structure describing what kind of events we subscribe to
+// For now, this is only on_process
+static const struct pw_stream_events stream_events = {
+    PW_VERSION_STREAM_EVENTS,
+    .process = on_process,
+};
+
+//  open a pipewire connection and initialize a non-blocking main loop
+//  pars[1] contains the name of the sink
+static AudioID *pipewire_open(void **pars)
+{
+    module_state *state = (module_state *)malloc(sizeof(module_state));
+    pw_init(0, NULL);
+    state->loop = pw_thread_loop_new("pipewire audio thread", NULL);
+    state->rb = SPA_RINGBUFFER_INIT();
+    state->sample_buffer = malloc(SAMPLE_BUFFER_SIZE);
+    // we don't know if the params array belongs to us
+    // it  could be freed by speech dispatcher at a later date, which means we could be storing a pointer to freed memory
+    // to ensure that the params array belongs to us and only we can free it, we will duplicate the string contained in params[1], even if such an operation would have not been necesary in the end
+    state->sink_name = strdup(pars[1]);
+    state->stream = pw_stream_new_simple(pw_thread_loop_get_loop(state->loop), state->sink_name, pw_properties_new(PW_KEY_MEDIA_TYPE, "Audio", PW_KEY_MEDIA_CATEGORY, "Playback", PW_KEY_MEDIA_ROLE, "Accessibility", NULL), &stream_events, &state);
+    return (AudioID *)state;
+}
+// configure a pipewire stream for the given speech dispatcher configuration and then prepair loop for playback
+static int pipewire_begin(AudioID *id, AudioTrack track)
+{
+    module_state *state = (module_state *)id;
+    const struct spa_pod *params[1];
+    uint8_t buffer[1024];
+    struct spa_pod_builder b = SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
+    int format;
+    state->channel_count = track.num_channels;
+    if (track.bits == 16)
+    {
+        switch (state->id.format)
+        {
+        case SPD_AUDIO_LE:
+            format = SPA_AUDIO_FORMAT_S16_LE;
+            break;
+        case SPD_AUDIO_BE:
+            format = SPA_AUDIO_FORMAT_S16_BE;
+            break;
+        default:
+            pw_log_error("invalid audio format specifier");
+            return -1;
+        }
+    }
+    else if (track.bits == 8)
+    {
+        format = SPA_AUDIO_FORMAT_S8;
+    }
+    else
+    {
+        pw_log_error("invalid audio format specifier");
+        return -1;
+    }
+
+    params[0] = spa_format_audio_raw_build(&b, SPA_PARAM_EnumFormat, &SPA_AUDIO_INFO_RAW_INIT(.format = format, .channels = track.num_channels, .rate = track.sample_rate));
+    pw_stream_connect(state->stream, PW_DIRECTION_OUTPUT, PW_ID_ANY, PW_STREAM_FLAG_AUTOCONNECT | PW_STREAM_FLAG_MAP_BUFFERS | PW_STREAM_FLAG_RT_PROCESS, params, 1);
```

Comment:
We don't want to connect on each `begin` call, only on first call and when the parameters change (e.g. format difference between several voices)

Reply:
what do you mean precisely by "multiple times per process": what process are you talking about?

`open` will be called once per output module, either in the module process itself (if the module doesn't support audio through server), or in the spd process (if the module does support audio through server). So there can be several open in the spd process. But that is completely *fine*: the `module_state *` pointer is meant for you to put all you need in it. In practice you don't need to care that open is called several times.

`begin` is called several times, for each synthesis, to tell which parameters are being used (rate, bits, etc.), but not concurrently.

Really, the scenario is *simple*, and you *don't* need to care more than this:

- `open` is called once
- `begin` is called, `feed_sync_overlap` is called several times to provide the pieces of audio as they get synthesized., `end` is called
- another `begin` +  several `feed_sync_overlap` + `end`
- etc.
- `close`, eventually

And the *only* concurrent call that you might encounter is `stop`.

> unless a new instance to the pipewire core object is created as far as I know

Yes, sure, you'll have one core object per `module_state *`.

> I should store them as an array of streams

No you don't, you just store one in your `module_state *`, no more. If you notice in `begin` that the rate/etc. is different from the previous one, just disconnect from the pw stream, and reconnect with updated rate/etc. See how it's done in e.g. the pulse plugin.

> the different copies of the library are mapped in different parts of memory, so they can't touch each others module_state structs

They *don't* need to be in different process to avoid touching to each other's `module_state`. You are *passed* the `module_state` you are supposed to be using.

Reply:
yes, you just need to record what the params were last time, no biggie :)